documentation/projects/proposals/data_normalization/20240227-implementation_plan_catalog_data_cleaning.md

```diff
@@ -0,0 +1,198 @@
+in the Catalog (possibly by the discontinued CommonCrawl ingestion) which is
+nowadays responsible for validating the provider data. However, it entails a
+problem of wasting resources both in time, which continues to increase, and in
+the machines (CPU) it uses, which could easily be avoided making the changes
+permanent by saving them in the upstream database.
+
+[img-data-refresh]:
+  https://github.com/WordPress/openverse-catalog/blob/main/DAGs.md#image_data_refresh
+
+## Expected Outcomes
+
+<!-- List any succinct expected products from this implementation plan. -->
+
+- The catalog database (upstream) preserves the cleaned data results of the
+  current Ingestion Server's cleaning steps
+- The image Data Refresh process is simplified by reducing the cleaning steps
+  time to nearly zero (and optionally removing them).
+
+## Step-by-step plan
+
+The cleaning functions that the Ingestion Server applies are already implemented
+in the Catalog in the `MediaStore` class: see its `_tag_blacklisted` method
+(which probably should be renamed) and the [url utilities][url_utils] file. The
+only part that it's not there and can't be ported is the filtering of
+low-confidence tags, since provider scripts don't save an "accuracy" by tag.
+
+With this the plan then starts in the Ingestion Server with the following steps:
+
+1. [Save TSV files of cleaned data to AWS S3](#save-tsv-files-of-cleaned-data-to-aws-s3)
+1. [Make and run a batched update DAG for one-time cleanup](#make-and-run-a-batched-update-dag-for-one-time-cleanup)
+1. [Run an image Data Refresh to confirm cleaning time is reduced](#run-an-image-data-refresh-to-confirm-cleaning-time-is-reduced)
+
+[url_utils]:
+  https://github.com/WordPress/openverse/blob/a930ee0f1f116bac77cf56d1fb0923989613df6d/catalog/dags/common/urls.py
+
+## Step details
+
+### Save TSV files of cleaned data to AWS S3
+
+In a previous exploration, it was set to store TSV files of the cleaned data in
+the form of `<identifier> <cleaned_field>`, which can be used later to perform
+the updates efficiently in the catalog DB, which only had indexes for the
+`identifier` field. These files are saved to the disk of the Ingestion Server
+EC2 instances, and worked fine for files with URL corrections since this type of
+fields is relatively short, but became a problem when trying to save tags, as
+the file turned too large and filled up the disk, causing problems to the data
+refresh execution.
+
+The alternative is to upload TSV files to the Amazon Simple Storage Service
+(S3), creating a new bucket or using `openverse-catalog` with a subfolder. The
+benefit of using S3 buckets is that they have streaming capabilities and will
+allow us to read the files in chunks later if necessary for performance. The
+downside is that objects in S3 don't allow appending, so it may require to
+upload files with different part numbers or evaluate if the [multipart upload
+process][aws_mpu] will serve us here.
```

Comment:
Noting here that I've used the [`smart_open` library](https://github.com/piskvorky/smart_open) in the past to seamlessly "stream" larger files into S3 without having to write them to disk locally!

Reply:
I came across this library too but thought it was only for fetching files. It will be great if it allows us to upload the file directly to ASW! Thanks for mentioning it 😄 I'll check it out.